lib/ansible/modules/cloud/amazon/ec2_vol_facts.py

```diff
@@ -59,57 +60,65 @@
 import traceback
 
 try:
-    import boto.ec2
-    from boto.exception import BotoServerError
-    HAS_BOTO = True
+    from botocore.exceptions import ClientError
+    from botocore.exceptions import ProfileNotFound
 except ImportError:
-    HAS_BOTO = False
+    pass  # caught by imported HAS_BOTO3
 
 from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import connect_to_aws, ec2_argument_spec, get_aws_connection_info
+from ansible.module_utils.ec2 import connect_to_aws, ec2_argument_spec, get_aws_connection_info, boto3_conn, HAS_BOTO3, boto3_tag_list_to_ansible_dict
+from ansible.module_utils.ec2 import ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict
 from ansible.module_utils._text import to_native
 
 
-def get_volume_info(volume):
+def get_volume_info(volume, region):
 
-    attachment = volume.attach_data
+    attachment = volume["attachments"]
+    tags = {}
+    for tag in volume["tags"]:
+        tags[tag["key"]] = tag["value"]
 
     volume_info = {
-        'create_time': volume.create_time,
-        'id': volume.id,
-        'encrypted': volume.encrypted,
-        'iops': volume.iops,
-        'size': volume.size,
-        'snapshot_id': volume.snapshot_id,
-        'status': volume.status,
-        'type': volume.type,
-        'zone': volume.zone,
-        'region': volume.region.name,
+        'create_time': volume["create_time"],
+        'id': volume["volume_id"],
+        'encrypted': volume["encrypted"],
+        'iops': volume["iops"] if "iops" in volume else None,
+        'size': volume["size"],
+        'snapshot_id': volume["snapshot_id"],
+        'status': volume["state"],
+        'type': volume["volume_type"],
+        'zone': volume["availability_zone"],
+        'region': region,
         'attachment_set': {
-            'attach_time': attachment.attach_time,
-            'device': attachment.device,
-            'instance_id': attachment.instance_id,
-            'status': attachment.status
+            'attach_time': attachment[0]["attach_time"] if len(attachment) > 0 else None,
+            'device': attachment[0]["device"] if len(attachment) > 0 else None,
+            'instance_id': attachment[0]["instance_id"] if len(attachment) > 0 else None,
+            'status': attachment[0]["state"] if len(attachment) > 0 else None,
+            'delete_on_termination': attachment[0]["delete_on_termination"] if len(attachment) > 0 else None
         },
-        'tags': volume.tags
+        'tags': tags
     }
 
     return volume_info
 
 
-def list_ec2_volumes(connection, module):
+def list_ec2_volumes(connection, module, region):
 
-    filters = module.params.get("filters")
+    # Replace filter key underscores with dashes, for compatibility, except if we're dealing with tags
```

Comment:
It would be nice to have integration tests to make sure this module stays backward compatible. It's not a blocker, but there *should* be tests for features even for preview modules. https://github.com/ansible/community/blob/master/group-aws/integration.md#policy

Reply:
I'm working on an integration test but it's hard to compare everything so let me post results here:

current version:

```
    "volumes": [
        {
            "attachment_set": {
                "attach_time": "2018-02-12T12:43:22.000Z",
                "device": "/dev/sda1",
                "instance_id": "i-88888888888888885",
                "status": "attached"
            },
            "create_time": "2018-02-12T12:43:22.461Z",
            "encrypted": false,
            "id": "vol-88888888888888888",
            "iops": 100,
            "region": "us-west-2",
            "size": 8,
            "snapshot_id": "snap-88888888888888887",
            "status": "in-use",
            "tags": {
                "Name": "test",
            },
            "type": "gp2",
            "zone": "us-west-2c"
        }
]
```

new version:

```
    "volumes": [
        {
            "attachment_set": {
                "attach_time": "2018-02-12T12:43:22+00:00", 
                "delete_on_termination": true, 
                "device": "/dev/sda1", 
                "instance_id": "i-88888888888888885", 
                "status": "attached"
            }, 
            "create_time": "2018-02-12T12:43:22.461000+00:00", 
            "encrypted": false, 
            "id": "vol-88888888888888888", 
            "iops": 100, 
            "region": "us-west-2", 
            "size": 8, 
            "snapshot_id": "snap-88888888888888887", 
            "status": "in-use", 
            "tags": {
                "Name": "test", 
            }, 
            "type": "gp2", 
            "zone": "us-west-2c"
        },
]
```

So as you can see date format is a bit different, is it something that should be checked/fixed?